overlord/devicestate/remodel.go

```diff
@@ -238,6 +254,38 @@ func (rc *baseRemodelContext) SystemMode() string {
 	return rc.systemMode
 }
 
+func (rc *baseRemodelContext) setRecoverySystemLabel(label string) {
+	rc.recoverySystemLabel = label
+}
+
+func (rc *baseRemodelContext) postFinish() error {
+	if rc.model.Grade() == asserts.ModelGradeUnset {
+		// nothing special for non-UC20 systems
+		return nil
+	}
+	if rc.recoverySystemLabel == "" {
+		return fmt.Errorf("internal error: recovery system label is unset during remodel finish")
+	}
+	// for UC20 systems we need record the fact that a new model is used for
+	// booting and consider a new recovery system as as seeded
+	oldDeviceContext := rc.GroundContext()
+	newDeviceContext := &rc.groundDeviceContext
+	if err := boot.DeviceChange(oldDeviceContext, newDeviceContext); err != nil {
+		return fmt.Errorf("cannot switch device: %v", err)
```

Comment:
can we still run either with the old or new system a recover if we error here? I don't remember the details but we can recover only from the recovery system matching the current system, right? I suppose the question also applies inside DeviceChange between the reseals. We should probably add explicit checks in the our test that check what recovery system can be used for recover at given interesting points

Reply:
Once this errors out, the model in the mdoeenv should match the model which comes from the device context. In a simple failure mode, the current model == old model, so we can recover using the old recovery system, but it will not be possible to unseal the data using the new system (keys are resealed when cleaning up after an error and the model file is restored). However, the device in state has already been updated, so if we reboot at this point but before the task is done, next time this task runs the remodel context's GroundDeviceContext() will return the new model because it was already in the state before the reboot. In this scenario, when DeviceChange runs again, it will do the right thing (there's a test for this in model.go), but only the new system will be able to recover if things go south.